### PR TITLE
feat: update on twitter ads integration

### PR DESCRIPTION
**Pull request by aanshi07**
## What are the changes introduced in this PR?

Updated Twitter Ads integration to include two additional identifiers: ip_address and user_agent.

## What is the related Linear task?

Resolves INT-2704

Doc: https://developer.x.com/en/docs/x-ads-api/measurement/web-conversions/api-reference/conversions

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [x] My code follows the style guidelines of this project

- [x] **No breaking changes are being introduced.**

- [x] All related docs linked with the PR?

- [x] All changes manually tested?

- [x] Any documentation changes needed with this change?

- [x] Is the PR limited to 10 file changes?

- [x] Is the PR limited to one linear task?

- [x] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3805/test-report.html

File: src/v0/destinations/twitter_ads/transform.js [hunk 1]
```diff
@@ -131,6 +131,20 @@ function processTrack(message, metadata, destination) {
...
 
+  if (message.properties.ip_address) {
+    const ipAddress = message.properties.ip_address.trim();
+    if (ipAddress) {
+      identifiers.push({ ip_address: ipAddress });
+    }
+  }
+
+  if (message.properties.user_agent) {
+    const userAgent = message.properties.user_agent.trim();
```
Comment (yashasvibajpai) [inline src/v0/destinations/twitter_ads/transform.js:142]: One small testcase can be added to cover both the lines @aanshi07

Comment (shrouti1507) [reply src/v0/destinations/twitter_ads/transform.js:142]: why can't we use https://www.rudderstack.com/docs/event-spec/standard-events/common-fields/#contextual-fields:~:text=and%20type.-,ip,-String

File: src/v0/destinations/twitter_ads/transform.js [hunk 2]
```diff
@@ -131,6 +131,20 @@ function processTrack(message, metadata, destination) {
...
+    if (userAgent) {
+      identifiers.push({ user_agent: userAgent });
```
Comment (shrouti1507) [inline src/v0/destinations/twitter_ads/transform.js:144]: why can't we use https://www.rudderstack.com/docs/event-spec/standard-events/common-fields/#contextual-fields:~:text=from%20the%20SDK.-,userAgent,-String

File: src/v0/destinations/twitter_ads/transform.js [hunk 3]
```diff
@@ -131,6 +131,20 @@ function processTrack(message, metadata, destination) {
     identifiers.push({ twclid: message.properties.twclid });
   }
 
+  if (message.properties.ip_address) {
```
Comment (shrouti1507) [inline src/v0/destinations/twitter_ads/transform.js:134]: why can't we use https://www.rudderstack.com/docs/event-spec/standard-events/common-fields/#contextual-fields:~:text=and%20type.-,ip,-String

Comment (yashasvibajpai) [reply src/v0/destinations/twitter_ads/transform.js:134]: if we are mapping `ip_address` and `user_agent` from context, why are setting same values in properties too? @aanshi07

File: test/integrations/destinations/twitter_ads/processor/data.ts [hunk 4]
```diff
@@ -894,6 +898,233 @@ export const data = [
...
+            },
+            statusCode: 200,
+          },
+        ],
+      },
+    },
+  },
+  {
+    name: 'twitter_ads',
+    description: 'Test 7',
```
Comment (yashasvibajpai) [inline test/integrations/destinations/twitter_ads/processor/data.ts:1025]: please update the description with what the test wants to show

File: test/integrations/destinations/twitter_ads/processor/data.ts [hunk 5]
```diff
@@ -894,6 +898,233 @@ export const data = [
       },
     },
   },
+  {
+    name: 'twitter_ads',
+    description: 'Test 6',
```
Comment (yashasvibajpai) [inline test/integrations/destinations/twitter_ads/processor/data.ts:903]: please update the description with what the test wants to show

File: test/integrations/destinations/twitter_ads/processor/data.ts (hunk 4 above)
Comment (yashasvibajpai) [reply test/integrations/destinations/twitter_ads/processor/data.ts:1025]: can you pl check in all tests if description is there

File: src/v0/destinations/twitter_ads/transform.js (hunk 3 above)
Comment (aanshi07) [reply src/v0/destinations/twitter_ads/transform.js:134]: As per the doc and implementation identifiers are required to be passed in properties.

File: test/integrations/destinations/twitter_ads/processor/data.ts (hunk 4 above)
Comment (aanshi07) [reply test/integrations/destinations/twitter_ads/processor/data.ts:1025]: Okay, I'll update for all the test cases.

**Review by yashasvibajpai — APPROVED** [review]

**Review by manish339k — APPROVED** [review]

**Comment by devops-github-rudderstack** [comment]
This PR is considered to be stale. It has been open for 20 days with no further activity thus it is going to be closed in 7 days. To avoid such a case please consider removing the stale label manually or add a comment to the PR.

**Comment by devops-github-rudderstack** [comment]
This PR is considered to be stale. It has been open for 20 days with no further activity thus it is going to be closed in 7 days. To avoid such a case please consider removing the stale label manually or add a comment to the PR.

**Pull request closed without merging**